src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -41,17 +42,30 @@ public object GetService(Type serviceType)
 
         internal object CaptureDisposable(object service)
         {
-            Debug.Assert(!_disposed);
-
             _captureDisposableCallback?.Invoke(service);
 
             if (ReferenceEquals(this, service) || !(service is IDisposable || service is IAsyncDisposable))
             {
                 return service;
             }
 
-            lock (ResolvedServices)
+            lock (_lock)
             {
+                if (_disposed)
+                {
+                    if (service is IDisposable disposable)
+                    {
+                        disposable.Dispose();
+                    }
+                    else
+                    {
+                        // this synchronous code path could cause a deadlock if we were to block on DisposeAsync
+                        // instead, a fire-and-forget call may cause race conditions while disposing borrowed resources.
+                        // since this is a rare error case, we just take the fire-and-forget approach here
+                        Task.Run(() => ((IAsyncDisposable)service).DisposeAsync().AsTask().GetAwaiter().GetResult());
```

Comment:
cc: @stephentoub

Reply:
I don't understand the comment.

But separate from that, if the goal is fire-and- 
forget, then you don't want or need the .GetAwaiter().GetResult(). And the Task.Run shouldn't be necessary either.

Reply:
We're not doing fire and forget, we're trying to avoid a deadlock for sync contexts that are single threaded in this rare case when somebody disposes during construction *and* they only implemented `IAsyncDisposable`.

Reply:
> We're not doing fire and forget,

As written,  it is. Maybe the last paren is misplaced.

Reply:
@maryamariyan we should update the comment. We're not doing fire and forget but we're doing sync over async here in the rare case where an object only implements IAsyncDisposable we may end up starving the thread pool.

Reply:
OK.